packages/next/src/client/components/react-dev-overlay/internal/components/copy-button/index.tsx

```diff
@@ -1,53 +1,95 @@
-import { useState } from 'react'
+import { useActionState, useEffect } from 'react'
 
-enum CopyState {
-  Initial = 'initial',
-  Success = 'success',
-  Pending = 'pending',
-  Error = 'error',
-}
+type CopyState =
+  | {
+      state: 'initial'
+    }
+  | {
+      state: 'error'
+      error: unknown
+    }
+  | { state: 'success' }
 
 export function CopyButton({
-  label,
+  actionLabel,
   successLabel,
   content,
   ...props
 }: React.HTMLProps<HTMLButtonElement> & {
-  label: string
+  actionLabel: string
   successLabel: string
   content: string
 }) {
-  const [copyState, setCopyState] = useState(CopyState.Initial)
-  const isDisabled = copyState === CopyState.Error
-  const title = isDisabled
-    ? ''
-    : copyState === CopyState.Success
-      ? successLabel
-      : label
+  const [copyState, dispatch, isPending] = useActionState(
+    async (state: CopyState, action: 'reset' | 'copy'): Promise<CopyState> => {
+      if (action === 'reset') {
+        return { state: 'initial' }
+      }
+      if (action === 'copy') {
+        if (!navigator.clipboard) {
+          return {
+            state: 'error',
+            error: new Error(
+              'Copy to clipboard is not supported in this browser'
+            ),
+          }
+        }
+        try {
+          await navigator.clipboard.writeText(content)
+          return { state: 'success' }
+        } catch (error) {
+          return { state: 'error', error }
+        }
+      }
+      return state
+    },
+    {
+      state: 'initial',
+    }
+  )
+
+  const error = copyState.state === 'error' ? copyState.error : null
+  useEffect(() => {
+    if (error !== null) {
+      // Additional console.error to get the stack.
+      console.error(error)
+    }
+  }, [error])
+  useEffect(() => {
+    if (copyState.state === 'success') {
+      const timeoutId = setTimeout(() => {
+        dispatch('reset')
+      }, 2000)
+
+      return () => {
+        clearTimeout(timeoutId)
+      }
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps -- dispatch is stable I believe
```

Comment:
Seems fine to add it if it's stable? We should try and avoid this rule since it makes it hard to catch when there are actually missing dependencies

Reply:
Maybe add `dispatch` along with a TODO comment that references https://github.com/facebook/react/pull/29665?

Reply:
[`f0951c9` (#66375)](https://github.com/vercel/next.js/pull/66375/commits/f0951c925b443b4634754d16786bf9e095f962ab)